File: tools/engine.py
"""Minimal node-render engine for node JSONs (multi-story).

Nodes live in ``stories/<story>/nodes/``; point ``nodes_dir`` at the
story you want to play (default: ``stories/fog_town/nodes``, the 雾镇
script).

Rendering contract (used by tests/ and by any future UI):

* ``render(node_id, state) -> {"id", "narrative", "choices": [...]}``
  Returns the node narrative plus the choices currently visible under
  ``state``. Choices whose ``condition`` is not satisfied are hidden.
  Each rendered choice carries ``text`` and — when the author wrote one —
  the per-choice ``outcome`` text.

* ``choose(node_id, choice_text, state, visits=None) -> {"outcome", "next_node", "state"}``
  Resolves the choice: branch ``when``/``otherwise``, variable effects
  (with 0-100 / 0-40 clamps), the SAN=0 -> Node_End_4 and
  SAN<30 -> Node_3_3 engine rules, and returns the player-facing
  ``outcome`` text to render immediately after the choice (``None`` for
  choices that carry no outcome).

Loop guard: choices may declare ``loop_guard`` (schema in the story's
nodes/README.md, e.g. ``stories/fog_town/nodes/README.md``). ``visits`` is a caller-owned mutable dict counting
entries into nodes that carry guarded choices; the guard attaches a
``hint`` once the count passes ``hint_after`` and locks the choice
(``locked: true``, unselectable) once it reaches ``max``.

Variables: sanity(0-100), bond(0-100), clue_a/b/c(0-40), key(bool).
CLUE = min(100, clue_a + clue_b + clue_c).
"""
import glob
import json
import os

# 默认读取《雾镇》剧本的节点目录；新剧本请传 nodes_dir 覆盖
DEFAULT_NODES_DIR = os.path.join("stories", "fog_town", "nodes")


def _clamp(value, lo=0, hi=100):
    return max(lo, min(hi, value))


class Engine:
    def __init__(self, nodes_dir=DEFAULT_NODES_DIR):
        self.nodes = {}
        for path in sorted(glob.glob(os.path.join(nodes_dir, "node_*.json"))):
            with open(path, encoding="utf-8") as fh:
                data = json.load(fh)
            self.nodes[data["id"]] = data

    # ------------------------------------------------------------------ state
    @staticmethod
    def initial_state():
        return {
            "sanity": 70,
            "bond": 0,
            "clue_a": 0,
            "clue_b": 0,
            "clue_c": 0,
            "key": False,
        }

    @staticmethod
    def clue(state):
        return min(100, state["clue_a"] + state["clue_b"] + state["clue_c"])

    def _meets(self, state, cond):
        if not cond:
            return True
        if cond.get("key") is not None and bool(cond["key"]) != bool(state["key"]):
            return False
        if "san_min" in cond and state["sanity"] < cond["san_min"]:
            return False
        if "san_max" in cond and state["sanity"] > cond["san_max"]:
            return False
        if "bond_min" in cond and state["bond"] < cond["bond_min"]:
            return False
        if "bond_max" in cond and state["bond"] > cond["bond_max"]:
            return False
        clue = self.clue(state)
        if "clue_min" in cond and clue < cond["clue_min"]:
            return False
        if "clue_max" in cond and clue > cond["clue_max"]:
            return False
        return True

    def _apply(self, state, effect):
        st = dict(state)
        st["sanity"] = _clamp(st["sanity"] + effect.get("sanity", 0))
        st["bond"] = _clamp(st["bond"] + effect.get("bond", 0))
        st["clue_a"] = _clamp(st["clue_a"] + effect.get("clue_a", 0), 0, 40)
        st["clue_b"] = _clamp(st["clue_b"] + effect.get("clue_b", 0), 0, 40)
        st["clue_c"] = _clamp(st["clue_c"] + effect.get("clue_c", 0), 0, 40)
        if effect.get("key"):
            st["key"] = True
        return st

    # ---------------------------------------------------------------- render
    def render(self, node_id, state, visits=None):
        node = self.nodes[node_id]
        visits = visits if visits is not None else {}
        count = visits.get(node_id, 0)
        choices = []
        for choice in node.get("choices", []):
            if not self._meets(state, choice.get("condition")):
                continue
            guard = choice.get("loop_guard")
            if guard and count >= guard.get("max", 0):
                choices.append(
                    {
                        "text": choice["text"],
                        "locked": True,
                        "locked_note": guard.get("locked_note"),
                    }
                )
                continue
            item = {"text": choice["text"], "outcome": choice.get("outcome")}
            if guard and count >= guard.get("hint_after", 0):
                item["hint"] = guard.get("hint")
            choices.append(item)
        return {"id": node_id, "narrative": node["narrative"], "choices": choices}

    # ---------------------------------------------------------------- choose
    def choose(self, node_id, choice_text, state, visits=None):
        node = self.nodes[node_id]
        visits = visits if visits is not None else {}
        count = visits.get(node_id, 0)
        for choice in node.get("choices", []):
            if choice["text"] == choice_text:
                break
        else:
            raise KeyError(f"no choice '{choice_text}' in {node_id}")

        if not self._meets(state, choice.get("condition")):
            raise ValueError(
                f"choice '{choice_text}' is not visible at {node_id} "
                f"under state {state}"
            )
        guard = choice.get("loop_guard")
        if guard and count >= guard.get("max", 0):
            raise ValueError(
                f"choice '{choice_text}' is locked by loop_guard at {node_id}"
            )

        target = choice["target"]
        effect = dict(choice.get("effect", {}))
        branch = choice.get("branch")
        if branch:
            if self._meets(state, branch.get("when", {})):
                effect.update(branch.get("effect", {}))
                if branch.get("target"):
                    target = branch["target"]
            else:
                effect.update(branch.get("otherwise", {}))

        next_state = self._apply(state, effect)

        # engine rules from 02_node_map.md 附录 B:
        # SAN=0 -> immediate E4; SAN<30 -> forced into Node_3_3
        # (transitions into Node_3_3 / Node_3_4 / endings are exempt).
        if next_state["sanity"] == 0:
            target = "Node_End_4"
        elif (
            next_state["sanity"] < 30
            and not target.startswith("Node_End")
            and target not in ("Node_3_3", "Node_3_4")
        ):
            target = "Node_3_3"

        # count entries into any node that carries loop-guarded choices
        if any(c.get("loop_guard") for c in self.nodes.get(target, {}).get("choices", [])):
            visits[target] = visits.get(target, 0) + 1

        return {
            "outcome": choice.get("outcome"),
            "next_node": target,
            "state": next_state,
        }

File: src/engine/engine.ts
/* 互动引擎 —— 移植自 tools/engine.py（语义对齐：钳位/条件/loop_guard/SAN 规则）。
   引擎本身不绑定任何故事：构造时传入节点列表即可。
   变量说明（见 stories/charon/nodes/README.md）：
   - sanity   SAN 值 0-100，归零触发净舱线，<30 强制进入 SAN 濒危节点
   - bond     信任/绑定度 0-100（真相派绑定，低 = 离心/未绑定）
   - clue_a/b/c  三类线索 0-40（表面证据 / 0.7 秒深层指纹 / 假死与真相）
   - key      逃生舱密钥（布尔）
   CLUE = min(100, clue_a + clue_b + clue_c) */

import type {
  Choice,
  ChoiceCondition,
  ChoiceEffect,
  GameState,
  NodeDef,
  Visits,
} from "./types";

export interface RenderedChoice {
  text: string;
  outcome?: string;
  locked?: boolean;
  locked_note?: string;
  hint?: string;
}

export interface RenderResult {
  id: string;
  narrative: string;
  choices: RenderedChoice[];
}

export interface ChooseResult {
  outcome: string | null;
  nextNode: string;
  state: GameState;
}

const clamp = (v: number, lo = 0, hi = 100) => Math.max(lo, Math.min(hi, v));

export function initialState(): GameState {
  return { sanity: 70, bond: 0, clue_a: 0, clue_b: 0, clue_c: 0, key: false, flags: {} };
}

export function clueTotal(s: GameState): number {
  return Math.min(100, s.clue_a + s.clue_b + s.clue_c);
}

export function meets(state: GameState, cond?: ChoiceCondition): boolean {
  if (!cond) return true;
  if (cond.key !== undefined && Boolean(cond.key) !== state.key) return false;
  if (cond.san_min !== undefined && state.sanity < cond.san_min) return false;
  if (cond.san_max !== undefined && state.sanity > cond.san_max) return false;
  if (cond.bond_min !== undefined && state.bond < cond.bond_min) return false;
  if (cond.bond_max !== undefined && state.bond > cond.bond_max) return false;
  const c = clueTotal(state);
  if (cond.clue_min !== undefined && c < cond.clue_min) return false;
  if (cond.clue_max !== undefined && c > cond.clue_max) return false;
  if (cond.flags) {
    for (const [name, want] of Object.entries(cond.flags)) {
      if (Boolean(state.flags?.[name]) !== want) return false;
    }
  }
  return true;
}

function apply(state: GameState, effect: ChoiceEffect): GameState {
  const st: GameState = {
    sanity: clamp(state.sanity + (effect.sanity ?? 0)),
    bond: clamp(state.bond + (effect.bond ?? 0)),
    clue_a: clamp(state.clue_a + (effect.clue_a ?? 0), 0, 40),
    clue_b: clamp(state.clue_b + (effect.clue_b ?? 0), 0, 40),
    clue_c: clamp(state.clue_c + (effect.clue_c ?? 0), 0, 40),
    key: state.key,
    flags: { ...state.flags },
  };
  if (effect.key) st.key = true;
  return st;
}

/** 选正文：取第一个 when 满足的叙事变体；没有则用节点默认正文。 */
function narrativeOf(node: NodeDef, state: GameState): string {
  for (const v of node.variants ?? []) {
    if (!v.when || meets(state, v.when)) return v.narrative;
  }
  return node.narrative;
}

export class Engine {
  nodes: Record<string, NodeDef>;

  constructor(nodes: NodeDef[]) {
    this.nodes = Object.fromEntries(nodes.map((n) => [n.id, n]));
  }

  render(nodeId: string, state: GameState, visits: Visits = {}): RenderResult {
    const node = this.nodes[nodeId];
    const count = visits[nodeId] ?? 0;
    const choices: RenderedChoice[] = [];
    for (const choice of node.choices) {
      if (!meets(state, choice.condition)) continue;
      const guard = choice.loop_guard;
      if (guard && count >= guard.max) {
        choices.push({ text: choice.text, locked: true, locked_note: guard.locked_note });
        continue;
      }
      const item: RenderedChoice = { text: choice.text, outcome: choice.outcome };
      if (guard && count >= guard.hint_after) item.hint = guard.hint;
      choices.push(item);
    }
    return { id: nodeId, narrative: narrativeOf(node, state), choices };
  }

  choose(nodeId: string, choiceText: string, state: GameState, visits: Visits = {}): ChooseResult {
    const node = this.nodes[nodeId];
    const count = visits[nodeId] ?? 0;
    const choice: Choice | undefined = node.choices.find((c) => c.text === choiceText);
    if (!choice) throw new Error(`no choice "${choiceText}" in ${nodeId}`);
    if (!meets(state, choice.condition)) {
      throw new Error(`choice "${choiceText}" is not visible at ${nodeId} under current state`);
    }
    const guard = choice.loop_guard;
    if (guard && count >= guard.max) {
      throw new Error(`choice "${choiceText}" is locked by loop_guard at ${nodeId}`);
    }

    let target = choice.target;
    let effect: ChoiceEffect = { ...(choice.effect ?? {}) };
    if (choice.branch) {
      if (meets(state, choice.branch.when)) {
        effect = { ...effect, ...(choice.branch.effect ?? {}) };
        if (choice.branch.target) target = choice.branch.target;
      } else {
        effect = { ...effect, ...(choice.branch.otherwise ?? {}) };
      }
    }

    const next = apply(state, effect);
    if (choice.flags) {
      for (const flag of choice.flags) next.flags[flag] = true;
    }

    // 引擎规则（对齐 tools/engine.py）：
    // SAN=0 -> 立即进入净舱线；SAN<30 -> 强制进入 Node_3_3（结局/濒危节点除外）。
    if (next.sanity === 0) {
      target = "Node_End_4";
    } else if (
      next.sanity < 30 &&
      !target.startsWith("Node_End") &&
      target !== "Node_3_3" &&
      target !== "Node_3_4"
    ) {
      target = "Node_3_3";
    }

    // 记录进入「带 loop_guard 选项」的节点次数
    const targetNode = this.nodes[target];
    if (targetNode && targetNode.choices.some((c) => c.loop_guard)) {
      visits[target] = (visits[target] ?? 0) + 1;
    }

    return { outcome: choice.outcome ?? null, nextNode: target, state: next };
  }
}
